File: scripts/linters/no_mod_rs.rb
def no_mod_rs(root_dir, excludes = [])
  pattern = root_dir.join('crates', '*', 'src', '**', 'mod.rs').to_s
  errors = Dir.glob(pattern).sort.filter_map do |path|
    relative = Pathname.new(path).relative_path_from(root_dir).to_s
    next if excludes.include?(relative)

    relative
  end

  return true if errors.empty?

  puts 'Found `mod.rs` file(s). Use `src/<submodule>.rs` instead of `<submodule>/mod.rs`:'
  errors.each do |path|
    puts "  #{path}"
  end
  false
end

File: scripts/linters/src/Linters/NoModRs.php
<?php

declare(strict_types=1);

namespace Shirabe\Lint\Linters;

use Shirabe\Lint\Linter;
use Shirabe\Lint\Support\FileFinder;
use Shirabe\Lint\Support\Paths;

final class NoModRs implements Linter
{
    public function name(): string
    {
        return 'no_mod_rs';
    }

    public function failureIntro(): string
    {
        return 'Found `mod.rs` file(s). Use `src/<submodule>.rs` instead of `<submodule>/mod.rs`:';
    }

    public function check(string $rootDir, array $excludes): array
    {
        $errors = [];

        foreach (FileFinder::modRsFiles($rootDir) as $path) {
            $relative = Paths::relativeTo($rootDir, $path);
            if (in_array($relative, $excludes, true)) {
                continue;
            }

            $errors[] = $relative;
        }

        return $errors;
    }
}
